secrethub_wrapper.go

```diff
@@ -24,37 +24,61 @@ struct SecretVersion {
 	time CreatedAt;
 	char* Status;
 };
+
+struct Client {
+	int ID;
+};
 #include <stdbool.h>
+#include <stdlib.h>
 */
 import "C"
 
 import (
 	"encoding/json"
+	"github.com/secrethub/secrethub-go/pkg/secrethub"
 	"os"
 	"strings"
-
-	"github.com/secrethub/secrethub-go/pkg/secrethub"
+	"unsafe"
 )
 
-// Client creates and returns a new Go client to be used by xgo clients.
-func Client() (*secrethub.Client, error) {
+var clients = make(map[int]secrethub.ClientInterface)
+var nextClientID = 1;
+
+//export new_Client
+func new_Client(errMessage **C.char) *C.struct_Client{
 	options := []secrethub.ClientOption{
 		secrethub.WithAppInfo(&secrethub.AppInfo{
 			Name:    "secrethub-xgo",
 			Version: "0.1.0",
 		}),
 	}
-	return secrethub.NewClient(options...)
-}
-
-// Read retrieves a secret by its path.
-//export Read
-func Read(path *C.char, errMessage **C.char) C.struct_SecretVersion {
-	client, err := Client()
+	client, err := secrethub.NewClient(options...)
 	if err != nil {
 		*errMessage = C.CString(err.Error())
-		return C.struct_SecretVersion{}
+		return nil
 	}
+	cClient := (*C.struct_Client)(C.malloc(C.size_t(unsafe.Sizeof(C.struct_Client{}))))
+	cClient.ID = C.int(nextClientID)
+	clients[nextClientID] = client
+	nextClientID++
+	return cClient
+}
+
+//export delete_Client
+func delete_Client(cClient *C.struct_Client) {
+	delete(clients, int(cClient.ID))
+	C.free(unsafe.Pointer(cClient))
+}
+
+func GetGoClient(cClient *C.struct_Client) secrethub.ClientInterface {
+	return clients[int(cClient.ID)]
```

Comment:
Should we guard against a scenario where the key does not exist?

Reply:
This cannot really happen with the C# client unless you somehow corrupt some memory or the C# runtime's garbage collection is broken. Entries from the map are only removed when the C# runtime does garbage collection, at which point you shouldn't have any reference to the client object.

Reply:
Okay, that's nice!

And what happens if we'd use this code to compile an SDK for a language other than C#?

Reply:
Good question. In the case of languages with garbage collection will (probably) work the same way. In the case of other languages we might need to adjust this to hide the panic.

I'm reluctant to putting a nice error message here since getting to such a situation would require code that is so bad you wouldn't even expect an error, just a crash. To get an error here you'd need to access memory after it has been freed, in this case we're managing the memory ourselves so we can make a nice error message, but it would probably be very confusing for someone who uses the client without knowing how it works under the hood.